test/groovy/TerraformStepTests.groovy

```diff
@@ -0,0 +1,249 @@
+    // on a feature branch (not  a change request) triggered by a periodic code scan
+    binding.setProperty('currentBuild', new CurrentBuild(
+      'SUCCESS',
+      ['hudson.triggers.PeriodicFolderTrigger']
+    ))
+    addEnvVar('BRANCH_NAME', 'feat/terraform')
+    script.call()
+    printCallStack()
+
+    // Then we expect a successful build
+    assertJobStatusSuccess()
+
+    // With all the stages called
+    /** Staging **/
+    assertTrue(assertMethodCallContainsPattern('stage', 'Staging'))
+    assertTrue(assertMethodCallContainsPattern('stage', 'Validate Terraform for Staging Environment'))
+    assertTrue(assertMethodCallContainsPattern('stage', 'Test Terraform Project'))
+
+    /** Production **/
+    assertTrue(assertMethodCallContainsPattern('stage', 'Production'))
+    assertTrue(assertMethodCallContainsPattern('stage', 'Generate Terraform Plan'))
+    assertTrue(assertMethodCallContainsPattern('sh', 'make --directory=.shared-tools/terraform/ plan'))
+    assertTrue(assertMethodCallContainsPattern('archiveArtifacts', 'terraform-plan-for-humans.txt'))
+
+    // Default is the principal branch: not a PR, no notification
+    assertFalse(assertMethodCallContainsPattern('stage', 'Notify User on the PR'))
+    assertFalse(assertMethodCallContainsPattern('publishChecks', dummyBuildUrl + 'artifact/terraform-plan-for-humans.txt'))
+    // Despite being on the principal branch, default build trigger is "timer"
+    assertFalse(assertMethodCallContainsPattern('stage', 'Waiting for User Manual Approval'))
+    assertFalse(assertMethodCallContainsPattern('input', 'Should we apply these changes to production?'))
+    assertFalse(assertMethodCallContainsPattern('stage', 'Shipping Changes'))
+    assertFalse(assertMethodCallContainsPattern('sh', 'make --directory=.shared-tools/terraform/ deploy'))
+
+    // Ensure that drift-detection is disabled
+    assertFalse(assertMethodCallContainsPattern('withEnv','TF_CLI_ARGS_plan=-detailed-exitcode'))
+
+    // And a daily cron trigger for the job
+    assertFalse(assertMethodCallContainsPattern('pipelineTriggers', '@daily'))
+  }
+
+  @Test
+  void itRunSuccessfullyWithManualBuildCause() throws Exception {
+    def script = loadScript(scriptName)
+
+    // when calling the shared library global function with defaults
+    // with a user manually-trigger build on the main branch
+    binding.setProperty('currentBuild', new CurrentBuild(
+      'SUCCESS',
+      ['hudson.model.Cause$UserIdCause']
+    ))
+    script.call()
+    printCallStack()
+
+    // Then we expect a successful build
+    assertJobStatusSuccess()
+
+    /** Staging is not run when user manually triggers the build **/
+    assertFalse(assertMethodCallContainsPattern('stage', 'Staging'))
+    assertFalse(assertMethodCallContainsPattern('stage', 'Validate Terraform for Staging Environment'))
+    assertFalse(assertMethodCallContainsPattern('stage', 'Test Terraform Project'))
```

Comment:
WDYT about defining variables or consts to define once the expected strings instead of repeating them in each test?

Reply:
that is a nice suggestion, but I'm not willing to refactor the whole PR here with this change: plese feel free to PR it :)